In [7]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
%matplotlib inline

plt.rcParams["figure.figsize"] = (12,6)

# NumPy

**Задача 1**\
Сгенерировать массив из случайных положительных чисел произвольной длины. Вывести элементы, которые больше среднего арифметического.

In [8]:
n = np.random.randint(10, 20)
arr = np.random.randint(1, 100, size=n)
print("Сгенерированный массив:", arr)

Сгенерированный массив: [12  2 58 85  1 27 53 92 54 60 82 25 52 51 18 82 11]


In [9]:
avg = np.mean(arr)
print("Среднее арифметическое: {:.2f}".format(avg))

Среднее арифметическое: 45.00


In [10]:
greater_than_avg = arr[arr > avg]
print("Элементы, большие среднего арифметического:", greater_than_avg)

Элементы, большие среднего арифметического: [58 85 53 92 54 60 82 52 51 82]


**Задача 2**\
Сгенерировать массив из 20 случайных элементов, элементы которого лежат в интервале [0, 10].
+ Определить количество элементов, отличающихся от минимального на 5.
+ Найти среднее арифметическое всех четных элементов массива, стоящих на нечетных местах.

In [11]:
arr = np.random.randint(0, 11, size=20)
arr

array([2, 9, 4, 8, 1, 6, 7, 7, 5, 6, 1, 2, 0, 8, 2, 7, 1, 8, 0, 8])

In [12]:
min_value = np.min(arr)
target_value = min_value + 5
n = np.count_nonzero(arr == target_value)
print(f"Количество элементов, отличающихся от минимального ({min_value}) на 5 равно {n}")

Количество элементов, отличающихся от минимального (0) на 5 равно 1


In [13]:
elements_odd_positions = arr[::2]
even_elements = elements_odd_positions[elements_odd_positions % 2 == 0]

if even_elements.size > 0:
    mean_even = np.mean(even_elements)
    print("Четные элементы, стоящие на нечетных местах:")
    print(even_elements)
    print("Среднее арифметическое этих элементов: {:.2f}".format(mean_even))
else:
    print("Нет четных элементов на нечетных местах.")

Четные элементы, стоящие на нечетных местах:
[2 4 0 2 0]
Среднее арифметическое этих элементов: 1.60


**Задача 3**\
Сгенерировать матрицу A размером 5 на 5, элементы которой лежат в интервале [1, 10].
+ Возвести в квадрат максимальный элемент матрицы.
+ Все четные числа заменить на 0.

In [14]:
A = np.random.randint(1, 11, size=(5, 5))
A

array([[ 8,  5,  5,  8, 10],
       [ 9,  6,  8, 10, 10],
       [ 3,  4,  4,  1,  4],
       [ 5,  5,  9,  6,  3],
       [ 1,  3, 10,  4,  2]])

In [15]:
max_val = np.max(A)
A[A == max_val] = A[A == max_val] ** 2
print("Максимальный элемент матрицы:", max_val)

Максимальный элемент матрицы: 10


In [16]:
A[A % 2 == 0] = 0
print("Матрица после замены всех четных чисел на 0:", A)

Матрица после замены всех четных чисел на 0: [[0 5 5 0 0]
 [9 0 0 0 0]
 [3 0 0 1 0]
 [5 5 9 0 3]
 [1 3 0 0 0]]


**Задача 4**\
Сгенерировать матрицу B размером 5 на 5, элементы которой лежат в интервале [-5, 5]. Отрицательные элементы матрицы заменить на число отрицательных элементов в строке, в которой они находятся.

In [17]:
B = np.random.randint(-5, 6, size=(5, 5))
B

array([[-4,  0,  2,  0, -5],
       [-2,  5,  4, -4,  2],
       [ 2, -5, -4, -5, -2],
       [ 5, -5, -3, -3,  3],
       [ 4, -4, -5, -1, -5]])

In [18]:
for i in range(B.shape[0]):
    count_negatives = np.sum(B[i] < 0)
    B[i][B[i] < 0] = count_negatives

print("Матрица B после замены отрицательных элементов:")
print(B)

Матрица B после замены отрицательных элементов:
[[2 0 2 0 2]
 [2 5 4 2 2]
 [2 4 4 4 4]
 [5 3 3 3 3]
 [4 4 4 4 4]]


# Pandas

In [19]:
df = pd.read_csv("https://raw.githubusercontent.com/rolandmueller/titanic/main/titanic3.csv")

In [20]:
df.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   int64  
 1   survived   1309 non-null   int64  
 2   name       1309 non-null   object 
 3   sex        1309 non-null   object 
 4   age        1046 non-null   float64
 5   sibsp      1309 non-null   int64  
 6   parch      1309 non-null   int64  
 7   ticket     1309 non-null   object 
 8   fare       1308 non-null   float64
 9   cabin      295 non-null    object 
 10  embarked   1307 non-null   object 
 11  boat       486 non-null    object 
 12  body       121 non-null    float64
 13  home.dest  745 non-null    object 
dtypes: float64(3), int64(4), object(7)
memory usage: 143.3+ KB


**Задание 1**
+ Какой была максимальная цена билета? Медиана?
+ Какова доля выживших?
+ Сколько мужчин / женщин находилось на борту?

In [22]:
max_fare = df['fare'].max()
print(f"Максимальная цена билета - {max_fare}")

Максимальная цена билета - 512.3292


In [23]:
median_fare  = df['fare'].median()
print(f"Медианная цена билета - {median_fare}")

Медианная цена билета - 14.4542


In [24]:
survival_rate = df['survived'].mean()
print(f"Доля выживших - {survival_rate}")

Доля выживших - 0.3819709702062643


In [25]:
gender_counts = df['sex'].value_counts()
gender_counts

male      843
female    466
Name: sex, dtype: int64

На борту находилось 466 женщин и 843 мужчин

**Задание 2**
+ Выведите распределение переменной __pclass__ для мужчин и женщин.
    + Сколько мужчин было в первом классе?
    + Сколько женщин было в третьем классе?
      
+ Создайте новый признак __age_group__, который равен:
    + 1, если пассажир был моложе 30 лет;
    + 2, если пассажиру было от 30 до 60 лет;
    + 3, если пассажир был старше 60 лет.

In [26]:
pclass_distribution = df.groupby(['sex', 'pclass']).size().unstack()
print("Распределение по классам для мужчин и женщин:")
print(pclass_distribution)

Распределение по классам для мужчин и женщин:
pclass    1    2    3
sex                  
female  144  106  216
male    179  171  493


In [27]:
men_first_class = pclass_distribution.loc['male', 1]
print(f"Количество мужчин в первом классе: {men_first_class}")

Количество мужчин в первом классе: 179


In [28]:
women_third_class = pclass_distribution.loc['female', 3]
print(f"Количество женщин в третьем классе: {women_third_class}")

Количество женщин в третьем классе: 216


In [38]:
def age_group(data):
    if data['age'] < 30:
        return 1
    elif data['age'] > 60:
        return 3
    else:
        return 2

In [39]:
df['age_group'] = df.apply(age_group, axis=1)
df[['age', 'age_group']].head()

age  age_group
0  29.0000          1
1   0.9167          1
2   2.0000          1
3  30.0000          2
4  25.0000          1

**Задание 3**\
Найдите самое популярное имя среди пассажиров:
+ мужчин;
+ женщин.

In [40]:
most_common_male_name = df[df['sex'] == 'male']["name"].str.split(". ").str[-1].mode()[0]
print(f'Самое популярное мужское имя - {most_common_male_name}')

Самое популярное мужское имя - John


In [41]:
most_common_female_name = df[df['sex'] == 'female']["name"].str.split(". ").str[-1].mode()[0]
print(f'Самое популярное женское имя - {most_common_female_name}')

Самое популярное женское имя - Mary


**Задание 4**
+ Выведите данные о всех пассажирах, которые сели в Cherbourg (Embarked='C') и были старше 50 лет.
+ Выведите данные о всех женщинах младше 20 и старше 50 лет.
+ Сколько пассажиров старше 70 лет было на борту?

In [42]:
df[(df['embarked'] == 'C') & (df['age'] > 50)]

pclass  survived                                               name  \
9         1         0                            Artagaveytia, Mr. Ramon   
43        1         1    Bucknell, Mrs. William Robert (Emma Eliza Ward)   
48        1         1   Candee, Mrs. Edward (Helen Churchill Hungerford)   
50        1         1  Cardeza, Mrs. James Warburton Martinez (Charlo...   
78        1         1  Compton, Mrs. Alexander Taylor (Mary Eliza Ing...   
104       1         1                     Eustis, Miss. Elizabeth Mussey   
123       1         1                   Frolicher-Stehli, Mr. Maxmillian   
135       1         0                          Goldschmidt, Mr. George B   
136       1         1                          Gracie, Col. Archibald IV   
174       1         0                            Kent, Mr. Edward Austin   
186       1         1          Lindstrom, Mrs. Carl Johan (Sigrid Posse)   
192       1         1                               Lurette, Miss. Elise   
215       1         0                         Newell, Mr. Arthur Webster   
221       1         0                     Ostby, Mr. Engelhart Cornelius   
233       1         1      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)   
246       1         0                             Rothschild, Mr. Martin   
247       1         1     Rothschild, Mrs. Martin (Elizabeth L. Barrett)   
252       1         0                         Ryerson, Mr. Arthur Larned   
264       1         1                Simonius-Blumer, Col. Oberst Alfons   
267       1         0                            Smith, Mr. James Clinch   
276       1         0                      Spencer, Mr. William Augustus   
280       1         1                    Stengel, Mr. Charles Emil Henry   
282       1         1    Stephenson, Mrs. Walter Bertram (Martha Eustis)   
303       1         0                           Warren, Mr. Frank Manley   
304       1         1   Warren, Mrs. Frank Manley (Anna Sophia Atkinson)   
308       1         1              White, Mrs. John Stuart (Ella Holmes)   
316       1         0                        Williams, Mr. Charles Duane   

        sex   age  sibsp  parch    ticket      fare            cabin embarked  \
9      male  71.0      0      0  PC 17609   49.5042              NaN        C   
43   female  60.0      0      0     11813   76.2917              D15        C   
48   female  53.0      0      0  PC 17606   27.4458              NaN        C   
50   female  58.0      0      1  PC 17755  512.3292      B51 B53 B55        C   
78   female  64.0      0      2  PC 17756   83.1583              E45        C   
104  female  54.0      1      0     36947   78.2667              D20        C   
123    male  60.0      1      1     13567   79.2000              B41        C   
135    male  71.0      0      0  PC 17754   34.6542               A5        C   
136    male  53.0      0      0    113780   28.5000              C51        C   
174    male  58.0      0      0     11771   29.7000              B37        C   
186  female  55.0      0      0    112377   27.7208              NaN        C   
192  female  58.0      0      0  PC 17569  146.5208              B80        C   
215    male  58.0      0      2     35273  113.2750              D48        C   
221    male  65.0      0      1    113509   61.9792              B30        C   
233  female  56.0      0      1     11767   83.1583              C50        C   
246    male  55.0      1      0  PC 17603   59.4000              NaN        C   
247  female  54.0      1      0  PC 17603   59.4000              NaN        C   
252    male  61.0      1      3  PC 17608  262.3750  B57 B59 B63 B66        C   
264    male  56.0      0      0     13213   35.5000              A26        C   
267    male  56.0      0      0     17764   30.6958               A7        C   
276    male  57.0      1      0  PC 17569  146.5208              B78        C   
280    male  54.0      1      0     11778   55.4417             C116        C   
282  female  

In [43]:
df[(df['sex'] == 'female') & ((df['age'] < 20) | (df['age'] > 50))]

pclass  survived                                               name  \
2          1         0                       Allison, Miss. Helen Loraine   
6          1         1                  Andrews, Miss. Kornelia Theodosia   
8          1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
11         1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
27         1         1            Bishop, Mrs. Dickinson H (Helen Walton)   
...      ...       ...                                                ...   
1265       3         0                          Van Impe, Miss. Catharina   
1273       3         0                 Vander Planke, Miss. Augusta Maria   
1279       3         0               Vestrom, Miss. Hulda Amanda Adolfina   
1300       3         1            Yasbeck, Mrs. Antoni (Selini Alexander)   
1304       3         0                               Zabour, Miss. Hileni   

         sex   age  sibsp  parch    ticket      fare    cabin embarked boat  \
2     female   2.0      1      2    113781  151.5500  C22 C26        S  NaN   
6     female  63.0      1      0     13502   77.9583       D7        S   10   
8     female  53.0      2      0     11769   51.4792     C101        S    D   
11    female  18.0      1      0  PC 17757  227.5250  C62 C64        C    4   
27    female  19.0      1      0     11967   91.0792      B49        C    7   
...      ...   ...    ...    ...       ...       ...      ...      ...  ...   
1265  female  10.0      0      2    345773   24.1500      NaN        S  NaN   
1273  female  18.0      2      0    345764   18.0000      NaN        S  NaN   
1279  female  14.0      0      0    350406    7.8542      NaN        S  NaN   
1300  female  15.0      1      0      2659   14.4542      NaN        C  NaN   
1304  female  14.5      1      0      2665   14.4542      NaN        C  NaN   

       body                        home.dest  age_group  
2       NaN  Montreal, PQ / Chesterville, ON          1  
6       NaN                       Hudson, NY          3  
8       NaN              Bayside, Queens, NY          2  
11      NaN                     New York, NY          1  
27      NaN                     Dowagiac, MI          1  
...     ...                              ...        ...  
1265    NaN                              NaN          1  
1273    NaN                              NaN          1  
1279    NaN                              NaN          1  
1300    NaN                              NaN          1  
1304  328.0                              NaN          1  

[135 rows x 15 columns]

In [44]:
print(f"Число пассажиров старше 70 лет - {df[df['age'] > 70]['age'].count()}")

Число пассажиров старше 70 лет - 6


**Задание 5**
+ Посчитайте средний возраст среди мужчин и среди женщин.
+ Посчитайте доли выживших в различных возрастных группах (новый признак **AgeGroup**).
+ Найдите максимальную стоимость билета в каждом из ценовых классов.

In [45]:
df.groupby('sex')['age'].mean()

sex
female    28.687071
male      30.585233
Name: age, dtype: float64

Средний возраст женщин - 28.7 лет, мужчин - 30.6 лет

In [46]:
df.groupby('age_group')['survived'].mean()

age_group
1    0.409490
2    0.366337
3    0.242424
Name: survived, dtype: float64

Во первой возрастной группе (до 30 лет) выжило 40.9% людей, во второй (от 30 до 60 лет) - 36,6%, в третьей (старше 60 лет) - 24,2%

In [49]:
df.groupby('age_group')['fare'].max()

age_group
1    263.0000
2    512.3292
3    263.0000
Name: fare, dtype: float64

В первой возрастной группе (до 30 лет) максимальная цена билета - 263 у.е., во второй (от 30 до 60 лет) - 512.3 у.е., в третьей (старше 60 лет) - 263 у.е.

**Задание 6**\
Ответьте на предыдущие вопросы с помощью функции crosstab и метода pivot_table.

In [52]:
print("Средний возраст среди мужчин и женщин")
pd.crosstab(index=df["sex"], columns="Средний возраст", values=df["age"], aggfunc="mean")

Средний возраст среди мужчин и женщин


col_0   Средний возраст
sex                    
female        28.687071
male          30.585233

In [53]:
print("Средний возраст среди мужчин и женщин")
pd.pivot_table(data=df, index="sex", values="age", aggfunc="mean")

Средний возраст среди мужчин и женщин


age
sex              
female  28.687071
male    30.585233

In [54]:
print("Доля выживших в различных возрастных группах")
pd.crosstab(index=df["age_group"], columns="Доля выживших", values=df["survived"], aggfunc="mean")

Доля выживших в различных возрастных группах


col_0      Доля выживших
age_group               
1               0.409490
2               0.366337
3               0.242424

In [55]:
print("Доля выживших в различных возрастных группах")
pd.pivot_table(data=df, index="age_group", values="survived", aggfunc="mean")

Доля выживших в различных возрастных группах


survived
age_group          
1          0.409490
2          0.366337
3          0.242424

In [56]:
print("Максимальная стоимость билета в каждом классе")
pd.crosstab(index=df["age_group"], columns="Максимальная цена билета", values=df["fare"], aggfunc="max")

Максимальная стоимость билета в каждом классе


col_0      Максимальная цена билета
age_group                          
1                          263.0000
2                          512.3292
3                          263.0000

In [57]:
print("Максимальная стоимость билета в каждом классе")
pd.pivot_table(data=df, index="age_group", values="fare", aggfunc="max")

Максимальная стоимость билета в каждом классе


fare
age_group          
1          263.0000
2          512.3292
3          263.0000